File: packages/core/src/parser/__testdata__/sample.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include "synapse.h"

#define MAX_RETRIES 3
#define BUFFER_SIZE 4096
#define VERSION "1.0.0"

typedef enum {
    STATUS_ACTIVE,
    STATUS_IDLE,
    STATUS_STOPPED,
    STATUS_ERROR
} AgentStatus;

typedef struct {
    char *model;
    int max_tokens;
    double temperature;
} AgentConfig;

typedef struct Agent {
    char name[256];
    AgentConfig config;
    AgentStatus status;
    struct Agent *next;
} Agent;

static Agent *agent_list = NULL;
static int agent_count = 0;

Agent *agent_create(const char *name, AgentConfig *config) {
    Agent *agent = (Agent *)malloc(sizeof(Agent));
    if (!agent) return NULL;

    strncpy(agent->name, name, 255);
    agent->name[255] = '\0';
    agent->config = *config;
    agent->status = STATUS_IDLE;
    agent->next = NULL;

    agent_count++;
    return agent;
}

void agent_destroy(Agent *agent) {
    if (!agent) return;
    free(agent->config.model);
    free(agent);
    agent_count--;
}

int agent_process(Agent *agent, const char *message, char *output, size_t out_size) {
    if (!agent || !message) return -1;
    agent->status = STATUS_ACTIVE;
    snprintf(output, out_size, "Response to: %s", message);
    agent->status = STATUS_IDLE;
    return 0;
}

AgentStatus agent_get_status(const Agent *agent) {
    return agent ? agent->status : STATUS_ERROR;
}

static void cleanup_all(void) {
    Agent *curr = agent_list;
    while (curr) {
        Agent *next = curr->next;
        agent_destroy(curr);
        curr = next;
    }
}

// TODO: add thread safety
// FIXME: potential buffer overflow in agent_process

File: packages/core/src/parser/__testdata__/sample.m
#import <Foundation/Foundation.h>
#import "SynapseAgent.h"

#define MAX_RETRIES 3
#define DEFAULT_MODEL @"claude-opus-4-6"

static NSInteger agentCount = 0;

typedef NS_ENUM(NSInteger, AgentStatus) {
    AgentStatusActive,
    AgentStatusIdle,
    AgentStatusStopped,
    AgentStatusError
};

@protocol AgentProtocol <NSObject>
- (NSString *)processMessage:(NSString *)message error:(NSError **)error;
- (NSArray<NSString *> *)getTools;
@property (nonatomic, readonly) AgentStatus status;
@end

@interface AgentConfig : NSObject
@property (nonatomic, copy) NSString *model;
@property (nonatomic, assign) NSInteger maxTokens;
@property (nonatomic, assign) double temperature;
+ (instancetype)defaultConfig;
@end

@interface SynapseAgent : NSObject <AgentProtocol>
@property (nonatomic, copy, readonly) NSString *name;
@property (nonatomic, strong) AgentConfig *config;
@property (nonatomic, assign) AgentStatus status;
- (instancetype)initWithName:(NSString *)name config:(AgentConfig *)config;
+ (instancetype)agentWithName:(NSString *)name;
@end

@implementation AgentConfig
+ (instancetype)defaultConfig {
    AgentConfig *config = [[AgentConfig alloc] init];
    config.model = DEFAULT_MODEL;
    config.maxTokens = 4096;
    config.temperature = 0.7;
    return config;
}
@end

@implementation SynapseAgent

- (instancetype)initWithName:(NSString *)name config:(AgentConfig *)config {
    self = [super init];
    if (self) {
        _name = [name copy];
        _config = config ?: [AgentConfig defaultConfig];
        _status = AgentStatusIdle;
        agentCount++;
    }
    return self;
}

+ (instancetype)agentWithName:(NSString *)name {
    return [[self alloc] initWithName:name config:nil];
}

- (NSString *)processMessage:(NSString *)message error:(NSError **)error {
    if (message.length == 0) {
        if (error) *error = [NSError errorWithDomain:@"SynapseAgent" code:400 userInfo:nil];
        return nil;
    }
    self.status = AgentStatusActive;
    NSString *result = [self callModel:message];
    self.status = AgentStatusIdle;
    return result;
}

- (NSArray<NSString *> *)getTools {
    return @[@"search", @"read", @"write"];
}

- (NSString *)callModel:(NSString *)message {
    // TODO: implement actual API call
    return [NSString stringWithFormat:@"Response to: %@", message];
}

- (void)dealloc {
    agentCount--;
}

@end

// FIXME: add proper error domain
